internal/armada/server/lease.go

```diff
@@ -334,6 +306,236 @@ func (q *AggregatedQueueServer) StreamingLeaseJobs(stream api.AggregatedQueue_St
+		for p, rs := range r.ResourcesLeasedByPriority {
+			resourcesByPriority[p] = schedulerobjects.ResourceList{
+				Resources: make(map[string]resource.Quantity),
+			}
+			for t, q := range rs.Resources {
+				resourcesByPriority[p].Resources[t] = q.DeepCopy()
+			}
+		}
+		report := &schedulerobjects.QueueClusterResourceUsage{
+			Queue:               r.Name,
+			ExecutorId:          req.GetClusterLeasedReport().ClusterId,
+			ResourcesByPriority: resourcesByPriority,
+		}
+		usageByQueue[r.Name] = report
+	}
+
+	// Store the usage so that other clusters can make use of it
+	clusterUsageReport := q.createClusterUsageReport(usageByQueue, req.Pool)
+	err := q.usageRepository.UpdateClusterQueueResourceUsage(req.ClusterId, clusterUsageReport)
+	if err != nil {
+		return nil, err
+	}
+
+	// load the usage from all other executors
+	reportsByExecutor, err := q.usageRepository.GetClusterQueueResourceUsage()
+	if err != nil {
+		return nil, err
+	}
+
+	// Create an aggregated usage by queue over all clusters
+	aggregatedUsageByQueue := q.aggregateUsage(reportsByExecutor, req.Pool)
+
+	// Collect all allowed priorities.
+	priorities := make([]int32, 0)
+	if len(q.schedulingConfig.Preemption.PriorityClasses) > 0 {
+		for _, p := range q.schedulingConfig.Preemption.PriorityClasses {
+			priorities = append(priorities, p)
+		}
+	} else {
+		priorities = append(priorities, 0)
+	}
+
+	// Nodes to be considered by the scheduler.
+	nodes := make([]*schedulerobjects.Node, len(req.Nodes))
+	for i, nodeInfo := range req.Nodes {
+		nodes[i] = schedulerobjects.NewNodeFromNodeInfo(&nodeInfo, req.ClusterId, priorities)
+	}
+
+	// Map queue names to priority factor for all active queues, i.e.,
+	// all queues for which the jobs queue has not been deleted automatically by Redis.
+	queues, err := q.queueRepository.GetAllQueues()
+	if err != nil {
+		return nil, err
+	}
+	priorityFactorByQueue := make(map[string]float64)
+	apiQueues := make([]*api.Queue, len(queues))
+	for i, queue := range queues {
+		priorityFactorByQueue[queue.Name] = float64(queue.PriorityFactor)
+		apiQueues[i] = &api.Queue{Name: queue.Name}
+	}
+	activeQueues, err := q.jobRepository.FilterActiveQueues(apiQueues)
+	if err != nil {
+		return nil, err
+	}
+	priorityFactorByActiveQueue := make(map[string]float64)
+	for _, queue := range activeQueues {
+		priorityFactorByActiveQueue[queue.Name] = priorityFactorByQueue[queue.Name]
+	}
+
+	sched, err := scheduler.NewLegacyScheduler(
+		q.schedulingConfig,
+		req.ClusterId,
+		nodes,
+		q.jobRepository,
+		priorityFactorByActiveQueue,
+	)
+	if err != nil {
+		return nil, err
+	}
+	sched.SchedulingReportsRepository = q.SchedulingReportsRepository
+	sched.MinimumJobSize = req.MinimumJobSize
+
+	// Give Schedule() a 3 second shorter deadline than ctx,
+	// to give it a chance to finish up before ctx is cancelled.
+	if deadline, ok := ctx.Deadline(); ok {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithDeadline(ctx, deadline.Add(-3*time.Second))
```

Comment:
do we need to do some sanity checking here to see what ctx.Deadline is?  If it's less than 3 seconds then bad things  are going to happen.

Reply:
The call will just time out I think? There should be no harm.